algorithms/kernel/pca/transform/oneapi/cl_kernels/pca_transform_cl_kernels.cl

```diff
@@ -0,0 +1,110 @@
+
+/*
+//++
+//  Implementation of PCA transform OpenCL kernels.
+//--
+*/
+
+#ifndef __PCA_TRANSFORM_CL_KERNELS_CL__
+#define __PCA_TRANSFORM_CL_KERNELS_CL__
+
+#include <string.h>
+
+#define DECLARE_SOURCE(name, src) static const char* name = #src;
+
+DECLARE_SOURCE(pca_transform_cl_kernels,
+
+__kernel void computeInvSigmas(__global algorithmFPType* rawVariances,
+                               __global algorithmFPType* invSigmas)
+{
+    const int tid = get_global_id(0);
+
+    if (rawVariances[tid] != (algorithmFPType)0)
+    {
+        invSigmas[tid] = (algorithmFPType)1 / (algorithmFPType)sqrt(rawVariances[tid]);
+    }
+    else
+    {
+        invSigmas[tid] = (algorithmFPType)0;
+    }
+}
+
+__kernel void normalize(__global algorithmFPType* pCopyBlock,
+						__global algorithmFPType* pRawMeans,
+						__global algorithmFPType* pInvSigmas,
+						const uint numMeans,
+                        const uint numInvSigmas,
+                        const uint maxWorkItemsPerGroup,
+                        const uint numFeatures)
+{
+    const int tid = get_local_id(0);
+    const int gid = get_group_id(0);
+    const int glid = get_global_id(0);
+    const int numWorkItemsPerGroup = get_local_size(0);
+    const int gnum = get_num_groups(0);
+
+    if (numFeatures > maxWorkItemsPerGroup)
+    {
+        uint amount = (numFeatures *  gnum) / maxWorkItemsPerGroup;
+
+        for(uint i = 0; i < amount + 1; i++)
+        {
+            if (numMeans != (algorithmFPType)0)
+            {
+                if (glid + gnum * numWorkItemsPerGroup * i < numFeatures *  gnum)
+                {
+                    pCopyBlock[glid + gnum * numWorkItemsPerGroup * i] = pCopyBlock[glid + gnum * numWorkItemsPerGroup * i] - pRawMeans[(glid + gnum * numWorkItemsPerGroup * i) % numFeatures];
+                }
+            }
+            if (numInvSigmas != (algorithmFPType)0)
+            {
+                if (glid + gnum * numWorkItemsPerGroup * i < numFeatures *  gnum)
+                {
+                    pCopyBlock[glid + gnum * numWorkItemsPerGroup * i] = pCopyBlock[glid + gnum * numWorkItemsPerGroup * i] * pInvSigmas[(glid + gnum * numWorkItemsPerGroup * i) % numFeatures];
+                }
+            }
+        }
+    }
+    else
+    {
+        if (numMeans != (algorithmFPType)0)
+        {
+            pCopyBlock[gid * numWorkItemsPerGroup + tid] = pCopyBlock[gid * numWorkItemsPerGroup + tid] - pRawMeans[tid];
+        }
+        if (numInvSigmas != (algorithmFPType)0)
+        {
+            pCopyBlock[gid * numWorkItemsPerGroup + tid] = pCopyBlock[gid * numWorkItemsPerGroup + tid] * pInvSigmas[tid];
+        }
+    }
+}
+
+__kernel void whitening(__global algorithmFPType* pTransformedBlock, 
+                        __global algorithmFPType* pInvEigenValues) 
+{
+    const int tid = get_local_id(0);
+    const int gid = get_group_id(0);
+    const int numFeatures = get_local_size(0);
```

Comment:
The local size can be less than the number of features due to device restrictions. This situation is not handled